OpenRA.Mods.Common/Traits/World/ClusterManager.cs

```diff
@@ -0,0 +1,314 @@
+				.Where(x => x.ConnectedCells
+				.Any(y => oldCluster.Cells
+				.Contains(y)));
+			if (affectedConnections.Any())
+				UpdateTransientConnections(affectedConnections, oldCluster, newCluster);
+
+			clusters.Remove(oldCluster);
+		}
+
+		protected void RecalculateCluster(ClusterContents cluster, List<CPos> deletedCells)
+		{
+			var selectedCluster = cluster;
+			var toProcess = new Queue<CPos>();
+			var toCheck = new List<CPos>(cluster.Cells.Where(x => !deletedCells.Contains(x)));
+			toCheck.Do(x => toProcess.Enqueue(x));
+
+			while (toProcess.Count != 0)
+			{
+				var start = toProcess.Dequeue();
+				if (!toCheck.Contains(start))
+					continue;
+
+				var clusterQueue = new Queue<CPos>();
+				clusterQueue.Enqueue(start);
+
+				while (clusterQueue.Count != 0)
+				{
+					var n = clusterQueue.Dequeue();
+					if (!toCheck.Contains(n))
+						continue;
+
+					toCheck.Remove(n);
+					UpdateCell(n, selectedCluster);
+
+					CVec.Directions.Select(d => d + n)
+						.Where(d => toCheck.Contains(d))
+						.Do(d => clusterQueue.Enqueue(d));
+				}
+
+				// Don't create a new cluster if all cells are checked
+				if (toCheck.Count != 0)
+				{
+					if (selectedCluster == cluster)
+					{
+						cluster.Cells.RemoveAll(x => toCheck.Contains(x));
+					}
+
+					selectedCluster = new ClusterContents();
+					clusters.Add(selectedCluster);
+				}
+			}
+		}
+
+		protected virtual ushort BuildDomain()
+		{
+			ClusterContents cluster = null;
+
+			var visited = new CellLayer<bool>(map);
+
+			var toProcess = new Queue<CPos>();
+			toProcess.Enqueue(MPos.Zero.ToCPos(map));
+
+			// Flood-fill over each cluster.
+			while (toProcess.Count != 0)
+			{
+				var start = toProcess.Dequeue();
+
+				// Technically redundant with the check in the inner loop, but prevents
+				// ballooning the cluster amount.
+				if (visited[start])
+					continue;
+
+				var clusterQueue = new Queue<CPos>();
+				clusterQueue.Enqueue(start);
+
+				var currentTrue = ClusterCondition(start);
+				cluster = new ClusterContents();
+				clusters.Add(cluster);
+
+				// Add all contiguous cells to our cluster, and make a note of
+				// any non-contiguous cells for future clusters.
+				while (clusterQueue.Count != 0)
+				{
+					var n = clusterQueue.Dequeue();
+					if (visited[n])
+						continue;
+
+					// add cell to queue if cell does not belong in queue
+					if (currentTrue != ClusterCondition(n))
+					{
+						toProcess.Enqueue(n);
+						continue;
+					}
+
+					visited[n] = true;
+					if (currentTrue)
+					{
+						clusterCellLayer[n] = cluster;
+						cluster.Cells.Add(n);
+					}
+					else
+						clusterCellLayer[n] = emptyCluster;
+
+					// Don't crawl off the map, or add already-visited cells.
+					CVec.Directions.Select(d => d + n)
+						.Where(d => visited.Contains(d) && !visited[d])
+						.Do(d => clusterQueue.Enqueue(d));
+				}
+			}
+
+			return cluster.ClusterId;
+		}
+	}
+
+	class ClusterContents
```

Comment:
You could make it hold a reference to a type (i.e. `ClusterContents<T>`), which should cover the use case of holding cash?

Reply:
Yes, and in a similar theme to `ClusterCondition` you could provide a `protected abstract T ClusterState(IEnumerable<CPos> cells);` so the user can specify how to determine the value. It could be called whenever you update the cells in a cluster.

Or you could go more granular and have methods for addition, removal, etc.